File: Raven.Client.Java/src/main/java/net/ravendb/client/IDocumentQuery.java
package net.ravendb.client;

import java.util.List;
import java.util.Map;

import net.ravendb.abstractions.basic.Lazy;
import net.ravendb.abstractions.closure.Action1;
import net.ravendb.abstractions.data.Facet;
import net.ravendb.abstractions.data.FacetResults;
import net.ravendb.abstractions.data.IndexQuery;
import net.ravendb.abstractions.data.QueryResult;
import net.ravendb.abstractions.json.linq.RavenJToken;
import net.ravendb.client.spatial.SpatialCriteria;

import com.mysema.query.types.Path;


/**
 * A query against a Raven index
 *
 * @param <T>
 */
public interface IDocumentQuery<T> extends IDocumentQueryBase<T, IDocumentQuery<T>>, Iterable<T> {
  /**
   * Selects the specified fields directly from the index
   * @param projectionClass The class of the projection
   * @param fields The fields.
   * @return
   */
  public <TProjection> IDocumentQuery<TProjection> selectFields(Class<TProjection> projectionClass, String... fields);

  /**
   * Selects the specified fields directly from the index
   * @param projectionClass The class of the projection
   * @param fields
   * @param projections
   * @return
   */
  public <TProjection> IDocumentQuery<TProjection> selectFields(Class<TProjection> projectionClass, String[] fields, String[] projections);

  /**
   * Selects the projection fields directly from the index
   * @param projectionClass The class of the projection
   * @return
   */
  public <TProjection> IDocumentQuery<TProjection> selectFields(Class<TProjection> projectionClass);

  /**
   * Sets user defined inputs to the query
   * @param queryInputs
   */
  public void setQueryInputs(Map<String, RavenJToken> queryInputs);

  /**
   * Gets the query result
   * Execute the query the first time that this is called.
   * @return The query result.
   */
  public QueryResult getQueryResult();

  /**
   * Register the query as a lazy query in the session and return a lazy
   * instance that will evaluate the query only when needed
   * @return
   */
  public Lazy<List<T>> lazily();

  /**
   * Register the query as a lazy query in the session and return a lazy
   * instance that will evaluate the query only when needed.
   * Also provide a function to execute when the value is evaluated
   * @param onEval
   * @return
   */
  public Lazy<List<T>> lazily(Action1<List<T>> onEval);

  /**
   * Create the index query object for this query
   * @return
   */
  public IndexQuery getIndexQuery();

  public IDocumentQuery<T> spatial(Path<?> path, SpatialCriteria criteria);

  public IDocumentQuery<T> spatial(String name, SpatialCriteria criteria);

  public boolean isDistinct();

  /**
   * Get the facets as per the specified doc with the given start and pageSize
   * @param facetSetupDoc
   * @param facetStart
   * @param facetPageSize
   * @return
   */
  public FacetResults getFacets(String facetSetupDoc, int facetStart, Integer facetPageSize);

  /**
   * Get the facets as per the specified facets with the given start and pageSize
   * @param facets
   * @param facetStart
   * @param facetPageSize
   * @return
   */
  public FacetResults getFacets(List<Facet> facets, int facetStart, Integer facetPageSize);

  /**
   * Query the facets results for this query using the specified facet document with the given start and pageSize
   * @param facetSetupDoc
   * @param start
   * @param pageSize
   * @return
   */
  public Lazy<FacetResults> toFacetsLazy(String facetSetupDoc);

  /**
   * Query the facets results for this query using the specified facet document with the given start and pageSize
   * @param facetSetupDoc
   * @param start
   * @param pageSize
   * @return
   */
  public Lazy<FacetResults> toFacetsLazy(String facetSetupDoc, int start);

  /**
   * Query the facets results for this query using the specified facet document with the given start and pageSize
   * @param facetSetupDoc
   * @param start
   * @param pageSize
   * @return
   */
  public Lazy<FacetResults> toFacetsLazy(String facetSetupDoc, int start, Integer pageSize);

  /**
   * Query the facets results for this query using the specified list of facets with the given start and pageSize
   * @param facets
   * @param start
   * @param pageSize
   * @return
   */
  public Lazy<FacetResults> toFacetsLazy(List<Facet> facets);

  /**
   * Query the facets results for this query using the specified list of facets with the given start and pageSize
   * @param facets
   * @param start
   * @param pageSize
   * @return
   */
  public Lazy<FacetResults> toFacetsLazy(List<Facet> facets, int start);

  /**
   * Query the facets results for this query using the specified list of facets with the given start and pageSize
   * @param facets
   * @param start
   * @param pageSize
   * @return
   */
  public Lazy<FacetResults> toFacetsLazy(List<Facet> facets, int start, Integer pageSize);

  /**
   * Query the facets results for this query using the specified facet document with the given start and pageSize
   * @param facetSetupDoc
   * @param start
   * @param pageSize
   * @return
   */
  public FacetResults toFacets(String facetSetupDoc);

  /**
   * Query the facets results for this query using the specified facet document with the given start and pageSize
   * @param facetSetupDoc
   * @param start
   * @param pageSize
   * @return
   */
  public FacetResults toFacets(String facetSetupDoc, int start);

  /**
   * Query the facets results for this query using the specified facet document with the given start and pageSize
   * @param facetSetupDoc
   * @param start
   * @param pageSize
   * @return
   */
  public FacetResults toFacets(String facetSetupDoc, int start, Integer pageSize);

  /**
   * Query the facets results for this query using the specified list of facets with the given start and pageSize
   * @param facets
   * @param start
   * @param pageSize
   * @return
   */
  public FacetResults toFacets(List<Facet> facets);

  /**
   * Query the facets results for this query using the specified list of facets with the given start and pageSize
   * @param facets
   * @param start
   * @param pageSize
   * @return
   */
  public FacetResults toFacets(List<Facet> facets, int start);

  /**
   * Query the facets results for this query using the specified list of facets with the given start and pageSize
   * @param facets
   * @param start
   * @param pageSize
   * @return
   */
  public FacetResults toFacets(List<Facet> facets, int start, Integer pageSize);

  /**
   * Returns first result
   * @return
   */
  public T first();

  /**
   * Returns first result
   * @return
   */
  public T firstOrDefault();

  /**
   * Materialize query, executes request and returns with results
   * @return
   */
  public List<T> toList();

  /**
   * Returns single result
   * @return
   */
  public T single();

  /**
   * Returns single result
   * @return
   */
  public T singleOrDefault();

  /**
   * Returns if any entry matches query
   * @return
   */
  public boolean any();

}

File: src/Raven.Client/IDocumentQuery.cs
//-----------------------------------------------------------------------
// <copyright file="IDocumentQuery.cs" company="Hibernating Rhinos LTD">
//     Copyright (c) Hibernating Rhinos LTD. All rights reserved.
// </copyright>
//-----------------------------------------------------------------------

using System;
using System.Collections.Generic;
using System.Linq.Expressions;

using Raven.Abstractions.Data;
using Raven.Client.Data;
using Raven.Client.Indexes;
using Raven.Client.Spatial;
using Raven.Json.Linq;

namespace Raven.Client
{
    /// <summary>
    ///     A query against a Raven index
    /// </summary>
    public interface IDocumentQuery<T> : IEnumerable<T>, IDocumentQueryBase<T, IDocumentQuery<T>>
    {
        /// <summary>
        ///     Whatever we should apply distinct operation to the query on the server side
        /// </summary>
        bool IsDistinct { get; }

        /// <summary>
        ///     Gets the query result. Accessing this property for the first time will execute the query.
        /// </summary>
        QueryResult QueryResult { get; }

        /// <summary>
        ///     Register the query as a lazy-count query in the session and return a lazy
        ///     instance that will evaluate the query only when needed.
        /// </summary>
        Lazy<int> CountLazily();

        /// <summary>
        ///     Get the facets as per the specified doc with the given start and pageSize
        /// </summary>
        /// <param name="queryInputs"></param>
        [Obsolete("Use SetTransformerParameters instead.")]
        IDocumentQuery<T> SetQueryInputs(Dictionary<string, RavenJToken> queryInputs);

        /// <summary>
        ///     Get the facets as per the specified facets with the given start and pageSize
        /// </summary>
        /// <param name="transformerParameters"></param>
        IDocumentQuery<T> SetTransformerParameters(Dictionary<string, RavenJToken> transformerParameters);

        /// <summary>
        ///     Create the index query object for this query
        /// </summary>
        IndexQuery GetIndexQuery(bool isAsync);

        /// <summary>
        ///     Register the query as a lazy query in the session and return a lazy
        ///     instance that will evaluate the query only when needed
        /// </summary>
        Lazy<IEnumerable<T>> Lazily();

        /// <summary>
        ///     Register the query as a lazy query in the session and return a lazy
        ///     instance that will evaluate the query only when needed.
        ///     Also provide a function to execute when the value is evaluated
        /// </summary>
        Lazy<IEnumerable<T>> Lazily(Action<IEnumerable<T>> onEval);

        /// <summary>
        ///     Selects the specified fields directly from the index if the are stored. If the field is not stored in index, value
        ///     will come from document directly.
        /// </summary>
        /// <typeparam name="TProjection">Type of the projection.</typeparam>
        /// <param name="fields">Array of fields to load.</param>
        IDocumentQuery<TProjection> SelectFields<TProjection>(params string[] fields);

        /// <summary>
        ///     Selects the specified fields directly from the index if the are stored. If the field is not stored in index, value
        ///     will come from document directly.
        /// </summary>
        /// <typeparam name="TProjection">Type of the projection.</typeparam>
        /// <param name="fields">Array of fields to load.</param>
        /// <param name="projections">Array of field projections.</param>
        IDocumentQuery<TProjection> SelectFields<TProjection>(string[] fields, string[] projections);

        /// <summary>
        ///     Selects the specified fields directly from the index if the are stored. If the field is not stored in index, value
        ///     will come from document directly.
        ///     <para>Array of fields will be taken from TProjection</para>
        /// </summary>
        /// <typeparam name="TProjection">Type of the projection from which fields will be taken.</typeparam>
        IDocumentQuery<TProjection> SelectFields<TProjection>();
        
        /// <summary>
        ///     Ability to use one factory to determine spatial shape that will be used in query.
        /// </summary>
        /// <param name="path">Spatial field name.</param>
        /// <param name="clause">function with spatial criteria factory</param>
        IDocumentQuery<T> Spatial(Expression<Func<T, object>> path, Func<SpatialCriteriaFactory, SpatialCriteria> clause);

        /// <summary>
        ///     Ability to use one factory to determine spatial shape that will be used in query.
        /// </summary>
        /// <param name="fieldName">Spatial field name.</param>
        /// <param name="clause">function with spatial criteria factory</param>
        IDocumentQuery<T> Spatial(string fieldName, Func<SpatialCriteriaFactory, SpatialCriteria> clause);

        /// <summary>
        ///     Sets a transformer to use after executing a query
        /// </summary>
        /// <param name="resultsTransformer"></param>
        IDocumentQuery<TTransformerResult> SetResultTransformer<TTransformer, TTransformerResult>()
            where TTransformer : AbstractTransformerCreationTask, new();

    }
}
